### PR TITLE
Disallow cookie auth for cors requests

### PR DESCRIPTION
**Pull request by BernhardPosselt**
@LukasReschke @DeepDiver1975 @MorrisJobke @PVince81

### CODE REVIEW

File: lib/private/appframework/http/request.php [hunk 1]
```diff
@@ -364,6 +364,14 @@ public function getCookie($key) {
 	}
 
 	/**
+	 * Shortcut for getting all cookie variables
+	 * @return array array equal to $_COOKIE
+	 */
+	public function getCookies() {
+		return $this->cookies;
+	}
```
Comment (LukasReschke) [inline lib/private/appframework/http/request.php]: Please add to public interface as well

File: lib/private/appframework/middleware/security/corsmiddleware.php [hunk 2]
```diff
@@ -50,6 +50,26 @@ public function __construct(IRequest $request,
...
+		// ensure that @CORS annotated API routes are not used in conjunction
+		// with session authentication since this enables CSRF attack vectors
+		if ($this->reflector->hasAnnotation('CORS') &&
+			count($this->request->getCookies()) > 0)  {
+			$msg = 'Session authentication is not allowed when using CORS';
+			throw new SecurityException($msg);
+		}
+	}
+
+
```
Comment (LukasReschke) [inline lib/private/appframework/middleware/security/corsmiddleware.php]: 1 new line too much

**Comment by Raydiation** [comment]
Ok, News android app explodes haha. sec

File: lib/private/appframework/middleware/security/corsmiddleware.php [hunk 3]
```diff
@@ -28,24 +28,52 @@
...
 
 	/**
 	 * @param IRequest $request
 	 * @param ControllerMethodReflector $reflector
 	 */
-	public function __construct(IRequest $request, 
-	                            ControllerMethodReflector $reflector) {
+	public function __construct(IRequest $request,
+	                            ControllerMethodReflector $reflector,
+	                            IUserSession $session) {
```
Comment (LukasReschke) [inline lib/private/appframework/middleware/security/corsmiddleware.php]: Function signatur needs to get fixed as well

Comment (Raydiation) [inline lib/private/appframework/middleware/security/corsmiddleware.php]: yes fixed

File: lib/private/appframework/middleware/security/corsmiddleware.php [hunk 4]
```diff
@@ -24,30 +24,57 @@
...
 class CORSMiddleware extends Middleware {
 
 	private $request;
 	private $reflector;
+	private $session;
 
 	/**
 	 * @param IRequest $request
 	 * @param ControllerMethodReflector $reflector
 	 */
```
Comment (LukasReschke) [inline lib/private/appframework/middleware/security/corsmiddleware.php]: PHPDoc? :smile:

**Comment by Raydiation** [comment]
News Android app works again with the PR \o/

**Comment by LukasReschke** [comment]
As discussed :+1:

Ref https://github.com/owncloud/core/issues/15894

**Comment by Raydiation** [comment]
Consider this PR a quick fix which will be made obsolete once the API stuff in #15894 is in

**Comment by LukasReschke** [comment]
(ah and the commits obviously needs squashing)

**Comment by Raydiation** [comment]
Ok, just tested in browser locally that this prevents CSRF :)

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by PVince81** [comment]
Makes sense :+1: 

Tags API still works (I see it was touched)

File: lib/private/appframework/middleware/security/corsmiddleware.php [hunk 5]
```diff
@@ -24,30 +24,69 @@
...
 	}
 
+	/**
+	 * This is being run in normal order before the controller is being
+	 * called which allows several modifications and checks
+	 *
+	 * @param Controller $controller the controller that is being called
+	 * @param string $methodName the name of the method that will be called on
+	 *                           the controller
+	 * @since 6.0.0
```
Comment (MorrisJobke) [inline lib/private/appframework/middleware/security/corsmiddleware.php]: Since 6.0.0?

**Comment by DeepDiver1975** [comment]
This will break cors api in shibboleth environments.  No idea how critical this is - but we need to at least document this.

**Comment by LukasReschke** [comment]
CORS will never have worked with Shibboleth before when invoked from another domain since we have the no credentials header. Thus a non-issue as far I can tell. 

But should be documented yes - will do later. That said CORS is only used by third-party apps atm.

**Comment by Raydiation** [comment]
The question is how we should handle this because it's a security fix that should be back ported. But apparently it will break some customers code

**Comment by Raydiation** [comment]
Criticality is that all api routes that use post and get are vulnerable to csrf

**Comment by Raydiation** [comment]
Api routes without cors annotation and with nocsrfrequired are still vulnerable BTW

**Comment by LukasReschke** [comment]
@Raydiation Please open backport PRs against 7+8 - we can discuss this then there. - Currently stable branches are frozen though.

cc @butonic You might need to have to adjust some code in owncloud/consulting

**Comment by Raydiation** [comment]
PRs open

**Pull request merged**